include/deal.II/matrix_free/operators.h

```diff
@@ -246,6 +246,86 @@ namespace MatrixFreeOperators
 
 
   /**
+   * Auxiliary class to provide interface vmult/Tvmult methods required in
+   * adaptive geometric multgrids. @p OperatorType class should be derived
+   * from MatrixFreeOperators::Base class.
+   *
+   * The adaptive multigrid realization in deal.II implements an approach
+   * called local smoothing. This means that the smoothing on the finest level
+   * only covers the local part of the mesh defined by the fixed (finest) grid
+   * level and ignores parts of the computational domain where the terminal
+   * cells are coarser than this level. As the method progresses to coarser
+   * levels, more and more of the global mesh will be covered. At some coarser
+   * level, the whole mesh will be covered. Since all level matrices in the
+   * multigrid method cover a single level in the mesh, no hanging nodes
+   * appear on the level matrices. At the interface between multigrid levels,
+   * homogeneous Dirichlet boundary conditions are set while smoothing. When
+   * the residual is transferred to the next coarser level, however, the
+   * coupling over the multigrid interface needs to be taken into account.
+   * This is done by the so-called interface (or edge) matrices that compute
+   * the part of the residual that is missed by the level matrix with
+   * homogeneous Dirichlet conditions. We refer to the @ref mg_paper
+   * "Multigrid paper by Janssen and Kanschat" for more details.
+   *
+   * For the implementation of those interface matrices, most infrastructure
+   * is already in place and provided by MatrixFreeOperators::Base through the
+   * two multiplication routines vmult_interface_down() and
+   * vmult_interface_up(). The only thing MGInterfaceOperator does is
+   * wrapping those operations and make them accessible via
+   * @p vmult() and @p Tvmult interface as expected by the multigrid routines
+   * (that were originally written for matrices, hence expecting those names).
+   * Note that the vmult_interface_down is used during the restriction phase of
+   * the multigrid V-cycle, whereas vmult_interface_up is used during the
+   * prolongation phase.
+   *
+   * @author Martin Kronbichler, 2016
```

Comment:
@kronbichler i put you as an author as I copy-paste whole lot of discussion from WIP step-37 here, otherwise the class is trivial.